### ISSUE TITLE
Copy remote directory with file module

### ISSUE THREAD

**Issue by ghost** [saltstack/salt#36762]
I'm trying to recursively copy a directory on the minion to another directory on the same minion, but it's not clear from the documentation for the file module (https://docs.saltstack.com/en/develop/ref/states/all/salt.states.file.html#salt.states.file.copy) whether this is possible.

The command I want to execute on the minion is:

```
cp -r /tmp/jre1.8.0_102 /usr/share/java/java-1.8.0-102
```

but unfortunately this state fails:

```
java_install:
  file.copy:
    - name: /usr/share/java/java-1.{{ java.version }}.0-{{ java['update'] }}
    - source: /tmp/jre1.{{ java.version }}.0_{{ java['update'] }}
    - recurse: True
```

with:

```
      ID: java_install
Function: file.copy
    Name: /usr/share/java/java-1.8.0-102
  Result: False
 Comment: Failed to copy "/tmp/jre1.8.0_102" to "/usr/share/java/java-1.8.0-102"
 Started: 10:30:36.660158
Duration: 1.656 ms
 Changes:
```

Can this be done with the file module or do I need to use `cmd.run`?

**Comment by rdutch** [comment]
For the source it is looking on the saltmaster, if you want to copy files locally on the minion I would use cmd.run.

**Comment by Ch3LL** [comment]
cmd.run is definitely an option but I think it does copy from the minion actually reaing the code an trying the following works:

```
sudo salt-call --local file.copy /tmp/blah1/blah1txt /tmp/blah2/blahtestcopy
```

Although that was a localclient call. @grobinson-blockchain What is in your debug logs on the minion when attempting to run this state file?

**Comment by boltronics** [comment]
The `file.copy` state module does actually copy *recursively* from another location on the same minion, but unlike the file.copy execution module, there is no `- recurse` argument - but it does recursively copy by default if a directory is provided. I had to look through the source to verify.

eg. This works:
```
Populate /var/www:
  file.copy:
    - name: /var/www
    - source: /srv/www/mywebsite
    - makedirs: True
    - preserve: True
    - subdir: True
```

I agree it's confusing and inconsistent. It doesn't help matters that `file.recurse` has a name that sounds closer to what we actually want - I had overlooked `file.copy` completely initially.

To address this issue correctly, I feel that ideally the state module should be updated to be more consistent with the execution module, or at the very least the documentation for the state module should clarify the behaviour.
